src/nnet3/attention.cc

```diff
@@ -0,0 +1,203 @@
+                          CuMatrixBase<BaseFloat> *C,
+                          Real beta) {
+  KALDI_ASSERT(A.NumCols() == B.NumCols() &&
+               A.NumRows() == C->NumRows());
+  int32 num_output_rows = A.NumRows(),
+      input_num_cols = A.NumCols(),
+      num_extra_rows = B.NumRows() - A.NumRows(),
+      context_dim = C->NumCols();
+  KALDI_ASSERT(num_extra_rows > 0 && num_extra_rows % (context_dim - 1) == 0);
+  int32 row_shift = num_extra_rows / (context_dim - 1);
+  CuMatrix<BaseFloat> Ctrans(C->NumCols(),
+                             C->NumRows());
+  for (int32 o = 0; o < context_dim; o++) {
+    CuSubVector<BaseFloat> c_col(Ctrans, o);
+    CuSubMatrix<BaseFloat> B_part(B, o * row_shift, num_output_rows,
+                                  0, input_num_cols);
+    c_col.AddDiagMatMat(alpha, A, kNoTrans, B, kTrans, 0.0);
+  }
+  C->CopyFromMat(Ctrans, kTrans);
+}
+
+
+void AttentionForward(Real key_scale,
+                      const CuMatrixBase<BaseFloat> &keys,
+                      const CuMatrixBase<BaseFloat> &queries,
+                      const CuMatrixBase<BaseFloat> &values,
+                      CuMatrixBase<BaseFloat> *c,
+                      CuMatrixBase<BaseFloat> *output) {
+  // First check the dimensions and values.
+  KALDI_ASSERT(key_scale > 0.0);
+  int32 num_input_rows = keys.NumRows(),
+      key_dim = keys.NumCols(),
+      num_output_rows = queries.NumRows(),
+      context_dim = queries.NumCols() - key_dim,
+      value_dim = values.NumCols();
+  KALDI_ASSERT(num_input_rows > 0 && key_dim > 0 &&
+               num_output_rows > num_input_rows &&
+               context_dim > 0 &&
+               (num_output_rows - num_input_rows) % (context_dim - 1) == 0 &&
+               values.NumRows() == num_output_rows);
+  KALDI_ASSERT(c->NumRows() == num_output_rows &&
+               c->NumCols() == context_dim);
+  KALDI_ASSERT(output->NumRows() == num_output_rows &&
+               (output->NumCols() == value_dim ||
+                output->NumCols() == value_dim + context_dim));
+
+  CuSubMatrix<BaseFloat> queries_key_part(
+      queries, 0, num_output_rows,
+      0, key_dim),
+      queries_context_part(
+          queries, 0, num_output_rows,
+          key_dim, context_dim);
+
+  GetAttentionDotProducts(key_scale,
+                          queries_key_part,
+                          keys, c);
+  // think of 'queries_context_part' as a position-dependent bias term.
+  c->AddMat(1.0, queries_context_part);
+  // compute the soft-max function.  Up till this point, 'c'
+  // actually contained what in attention.h we called 'b', which is
+  // the input to the softmax.
+  c->SoftMaxPerRow(*c);
+
+
+  // the part of the output that is weighted
+  // combinations of the input values.
+  CuSubMatrix<BaseFloat> output_values_part(
+      *output, 0, num_output_rows, 0, value_dim);
+
+  ApplyScalesToOutput(1.0, values, *c, &output_values_part);
```

Comment:
Since it is asserted (a few lines before) that `values` and `c` both have `num_output_rows` rows, this will set A, B, and C in a way that all have the same number of rows, so `row_shift` will become `0`.